[SimpleLoopUnswitch] Don't use BlockFrequencyInfo to skip cold loops

lukel97 opened the pull request:
In https://reviews.llvm.org/D129599, non-trivial switching was disabled for cold loops in the interest of code size. This added a dependency on BlockFrequencyInfo with PGO, but in loop passes this is only available on a lossy basis: see https://reviews.llvm.org/D86156

LICM moved away from BFI so as of today SimpleLoopUnswitch is the only remaining loop pass that uses BFI, for the sole reason to prevent code size increases in PGO builds. It doesn't use BFI if there's no profile summary available.

After some investigation on llvm-test-suite it turns out that the lossy BFI causes very significant deviations in block frequency, since when new loops are deleted/created during the loop pass manager it can return frequencies for different loops altogether. 
This results in unswitchable loops being mistakenly skipped because they are thought to be cold.

This patch removes the use of BFI from SimpleLoopUnswitch and thus the last remaining use of BFI in a loop pass. 

To recover the original intent of not unswitching cold code, PGOForceFunctionAttrs can be used to annotate functions which can be optimized for code size, since SimpleLoopUnswitch will respect OptSize: https://reviews.llvm.org/D94559

This isn't 100% the same behaviour since the previous behaviour checked for coldness at the loop level and this is now at the function level. We could expand PGOForceFunctionAttrs to be more granular at the loop level, https://github.com/llvm/llvm-project/issues/159595 tracks this idea.

lukel97 committed:
[SimpleLoopUnswitch] Don't use BlockFrequencyInfo to skip cold loops

In https://reviews.llvm.org/D129599, non-trivial switching was disabled for cold loops in the interest of code size. This added a dependency on BlockFrequencyInfo, but in loop passes this is only available on a lossy basis: see https://reviews.llvm.org/D86156

LICM moved away from BFI and as of today SimpleLoopUnswitch is the only remaining loop pass that uses BFI, for the sole reason to prevent code size increases in PGO builds. It doesn't use BFI if there's no profile summary available.

However just before the BFI check, we also check to see if the function is marked as OptSize: https://reviews.llvm.org/D94559

And coincidentally sometime after the initial BFI patch PGOForceFunctionAttrsPass was added which will automatically annotate cold functions with OptSize: https://reviews.llvm.org/D149800

I think using PGOForceFunctionAttrs to add the OptSize is probably a more accurate and generalized way to prevent unwanted code size increases. So this patch proposes to remove the BFI check in SimpleLoopUnswitch.

This isn't 100% the same behaviour since the previous behaviour checked for coldness at the loop level and this is now at the function level, but I believe the benefits outweigh this:

- It allows us to remove BFI from the function to loop pass adapter, which was only enabled for certain stages in the LTO pipeline
- We no longer have to worry about lossy analysis results
- Which in turn means the decision to avoid non-trivial switching will be more accurate
- It brings the behaviour inline with other passes that respect OptSize
- It respects the -pgo-cold-func-opt flag so users can control the behaviour
- It prevents the need to run OuterAnalysisManagerProxy as often which is probably good for compile time

lukel97 commented on llvm/test/Transforms/PhaseOrdering/unswitch-cold-func.ll:
I've moved the existing test to PhaseOrdering to show how using PGOForceFunctionAttrs also prevents unswitching this loop

lukel97 commented:
> This only works if the whole function is cold, not if there is a cold loop inside a hot function, right?

Yeah I briefly touched on that in the PR description. I hope that function level granularity is good enough for whatever case motivated the initial patch. At least it seems to be good enough for the test case that was included with it, [llvm/test/Transforms/PhaseOrdering/unswitch-cold-func.ll](https://github.com/llvm/llvm-project/pull/159522#diff-89d1a06d66e24dde89e8d3805d09485fe7691b7dab9d3d924624f252bb5cb7c5)

lukel97 committed:
Fix polly

nikic commented:
Okay, looking through the history of what happened here, we have:

 * https://github.com/llvm/llvm-project/commit/f756f06cc471b91d81d7f7f15d9df76ed24de730 initially checked whether the loop is cold
 * https://github.com/llvm/llvm-project/commit/fb45f3c9486f5d9e3003db95386432562b23577c switched this to check whether the function is cold (using isFunctionColdInCallGraph) to prevent SPEC regressions from the previous
 * https://github.com/llvm/llvm-project/commit/dfb40d3fd7a261d7e3f697242df4680f1e6780ff then switched to checking for a cold loop nest because of compile-time issues with the previous approach.

So with that history in mind, it does seem like just relying on the whole function being cold (via optsize) may be fine, as that was one of the intermediate solutions, which was only dropped due to compile-time issues. (Note that PGOForceFunctionAttrsPass uses isFunctionColdInCallGraph as well, so it should be equivalent to the second solution without the compile-time impact.)

lukel97 committed:
Remove preserve-bpi.ll test

This was added to illustrate the lossy preservation of BPI in 452714f8f8037ff37f9358317651d1652e231db2

But it fails now because we don't run OuterAnalysisManagerProxy beforehand. I'm not sure if it's still a useful thing to test given that #159516 removes BPI from the adaptor, so I've gone ahead and deleted the test

nikic reviewed (approved):
LGTM, but please leave some time for others to comment.

lukel97 commented on llvm/lib/Passes/PassBuilder.cpp:
```diff
@@ -2095,14 +2093,8 @@ Error PassBuilder::parseFunctionPass(FunctionPassManager &FPM,
-      });
-      bool UseBPI = llvm::any_of(InnerPipeline, [](auto Pipeline) {
-        return Pipeline.Name == "loop-predication";
-      });
```
Yeah, I think this change comes from the stacked PR https://github.com/llvm/llvm-project/pull/159516/files

fhahn reviewed (commented):
Overall checking for hot/cold loops in principle makes more sense to me; i.e. no point in unswitching a loop that's never executed in a hot function.

Not regressing SPEC is great, but still leaves a lot of room to regress other workloads. I don't really have any concrete suggestions at the moment; if we can't easily check for hotness of loops I guess this is fine for now and we can revisit if/when we discover regressions

teresajohnson commented:
> Overall checking for hot/cold loops in principle makes more sense to me; i.e. no point in unswitching a loop that's never executed in a hot function.

Agree, this change isn't going to be equivalent. I'm a little unclear of the problem being solved by this PR. What is the main motivation? There's a list of benefits but I'm not sure how important they are.

> Not regressing SPEC is great, but still leaves a lot of room to regress other workloads. I don't really have any concrete suggestions at the moment; if we can't easily check for hotness of loops I guess this is fine for now and we can revisit if/when we discover regressions

lukel97 commented:
> Agree, this change isn't going to be equivalent. I'm a little unclear of the problem being solved by this PR. What is the main motivation? There's a list of benefits but I'm not sure how important they are.

The main motivation is to remove the last use of BlockFrequencyInfo from within a loop pass, because I'm not sure if there's much value in relying on a potentially inaccurate result given that it's only preserved lossily.

I noticed this originally when some of the lossy results escaped into the loop vectorizer in #157888, causing BFI to return some bogus frequencies. Not off by one or two etc, but returning the frequencies for different blocks altogether since the CFG had been modified. I'm a bit worried that SimpleLoopUnswitch is also affected by this e.g. thinking a loop is cold when it's not. 

There is a secondary motivation that avoiding code size increases should be handled by PGOForceFunctionAttrsPass. I think it's somewhat coincidental that the patch checked that the loops were cold instead of the function. As Nikita mentions, fb45f3c9486f5d9e3003db95386432562b23577c seems to suggest that the original author was fine with checking the function.

lukel97 commented:
> Isn't this a larger problem though - downstream function passes such as BB layout will also be affected by incorrect BFI. I'm unclear why disabling use of BFI in this particular pass is the right approach to this issue.

Incorrect BFI leaking through to function passes was fixed in #157888 (wasn't exactly the easiest thing to debug 😅). AFAIK SimpleLoopUnswitch is now the last user of lossy BFI.

teresajohnson commented:
Thanks for the data. That does look pretty compelling. We may have some binary size or compile time increase for loops that still are cold in non-cold functions - do you have any data on how many cold loops (after forcing BFI/BPA invalidation before the pass as above) will be unnecessarily unswitched after this change due to being in non-cold functions?

Also can you check a couple of instances where the loop was cold until BFI/BPA were invalidated and see why they were cold - were they new loops created by previous loop transformations? With any eventual support for loop attributes we'd need to ensure they were copied or synthesized on such new loop nests appropriately.

But generally I think this change makes sense. However, I have a question on a test change - it isn't clear to me whether the effects of this PR are being captured in a test currently.

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into simpleloopunswitch-remove-bfi

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into simpleloopunswitch-remove-bfi

lukel97 committed:
Update function to reflect that it's not actually hot

teresajohnson reviewed (approved):
Thanks for digging into this, lgtm

lukel97 committed:
Merge branch 'main' into simpleloopunswitch-remove-bfi
